main/src/main/scala/sbt/Defaults.scala

```diff
@@ -2497,19 +2497,18 @@ object Defaults extends BuildCommon {
       val contents = AnalysisContents.create(analysisResult.analysis(), analysisResult.setup())
       store.set(contents)
       Def.declareOutput(analysisOut)
-      val dir = classDirectory.value
-      if (dir / "META-INF" / "MANIFEST.MF").exists then IO.delete(dir / "META-INF" / "MANIFEST.MF")
-      // inline mappings
-      val mappings = Path
-        .allSubpaths(dir)
-        .filter(_._1.isFile())
-        .map { case (p, path) =>
-          val vf = c.toVirtualFile(p.toPath())
-          (vf: HashedVirtualFileRef) -> path
-        }
-        .toSeq
+      val dir = classDirectory.value.toPath
+      val mappings =
+        for
+          ref <- analysisResult.analysis.readStamps.getAllProductStamps.asScala.keysIterator
+          path = c.toPath(ref)
+          if path.toFile().isFile() && path.startsWith(dir)
+        yield
+          val hashedRef: HashedVirtualFileRef = c.toVirtualFile(path)
+          val relativePath = dir.relativize(path).toString
+          hashedRef -> relativePath
```

Comment:
When traversing the class directory we get all the old resource files, which later messes up with the `copyResources` sync. To avoid that we use the list of products in the compilation analysis.